state/secretbackends.go

```diff
@@ -307,3 +353,243 @@ func (st *State) incBackendRevisionCountOps(backendID string) ([]txn.Op, error)
 	}
 	return []txn.Op{countOp, incOp}, nil
 }
+
+type secretBackendRotationDoc struct {
+	DocID    string `bson:"_id"`
+	TxnRevno int64  `bson:"txn-revno"`
+
+	// These fields are denormalised here so that the watcher
+	// only needs to access this collection.
+	Name           string    `bson:"backend-name"`
+	NextRotateTime time.Time `bson:"next-rotate-time"`
+}
+
+func (s *secretBackendsStorage) tokenRotationOps(ID string, name *string, nextRotateTime *time.Time) ([]txn.Op, error) {
+	if nextRotateTime == nil && name == nil {
+		return []txn.Op{{
+			C:      secretBackendsRotateC,
+			Id:     ID,
+			Remove: true,
+		}}, nil
+	}
+	secretBackendRotateCollection, closer := s.st.db().GetCollection(secretBackendsRotateC)
+	defer closer()
+
+	var doc secretBackendRotationDoc
+	err := secretBackendRotateCollection.FindId(ID).One(&doc)
+	if err == mgo.ErrNotFound {
+		return []txn.Op{{
```

Comment:
Is it possible name == nil or nextRotateTime == nil here？
If that's true, panic?

Reply:
the calling logic sets these to non - nil when inserting a new record